### DISCUSSION TITLE
Views with Security Definers View

### DISCUSSION THREAD

**Discussion by impossibleapp** [supabase/supabase#28464] [Questions] [answered]
I have created 3 views and all of them are showing up with the below error (rather than a warning). I understand that as these views have creators access rather than readers, it can lead to unauthorised access. However, I don't know how to change the permissions of a view and change the permissions so that it is secure.

Below is the error description:
Detects views defined with the SECURITY DEFINER property. These views enforce Postgres permissions and row level security policies (RLS) of the view creator, rather than that of the querying user

![image](https://github.com/user-attachments/assets/95475576-086d-4b15-9dfd-03220bac478b)

**Comment by GaryAustin1** [comment, marked as answer]
https://supabase.com/docs/guides/database/tables?queryGroups=database-method&database-method=sql&queryGroups=language&language=sql#view-security

**Reply by impossibleapp** [reply]
got the script. thanks!

-- alter a security_definer view to be security_invoker
alter view <view name>
set (security_invoker = true);

-- create a view with the security_invoker modifier
create view <view name> with(security_invoker=true) as (
  select * from <some table>
);

**Reply by gnewfield** [reply]
Quick callout that may save others some time (and is perhaps worth looking into by the Supabase team): I was only able to change these settings in my **remote** database. When I tried to execute these queries in the **local** Supabase Studio SQL editor, the query would execute without error. However, after running `supabase db diff`, it would tell me that no schema changes were found. After trying multiple different approaches to changing the view locally, I decided to modify the remote view directly using the prod web UI and sql editor. This worked and immediately resolved my security error.

**Reply by GaryAustin1** [reply]
![image](https://github.com/user-attachments/assets/03443025-b123-4c2c-8f80-376a96e3fbfd)
